# Assignment #4 - Data from the Web, Combining Data

## 1. Read the 2021 course schedule into a DataFrame

In [305]:
import pandas as pd
from bs4 import BeautifulSoup

In [306]:
#Read the whole html
cs2021 = BeautifulSoup(open("spr2021schedule.htm"), 'html.parser')
#The schedule information are all inside .schedule-listing class
schedulelisting = cs2021.select('.schedule-listing')[0]

#### Parse number-section into array

In [307]:
#Count number of 'li', which are lists of individual classes
numseccount = schedulelisting.select('li')
numbersectionarr = []
#Append all numbersection into array
for i in range(len(numseccount)):
    numbersectionarr.append(schedulelisting.select('li')[i]['id'])
#Clean into appropriate format
for i in range(len(numbersectionarr)):
    numbersectionarr[i] = numbersectionarr[i].upper()

#### Parse course name into array

In [308]:
#All course names are inside class .expand
coursename = schedulelisting.select('.expand')
#Array to store course names
coursenamearr = []
#Append while removing newlines
for i in range(len(coursename)):
    coursenamearr.append(coursename[i].get_text())
    coursenamearr[i] = " ".join(coursenamearr[i].strip().split())

#### Parse instructor name and time into array

In [309]:
#Since instructor and time columns share similar attributes with class col-sm 2 in HTML, parse and clean them simultaneously
instructorAndTime = schedulelisting.select('.col-sm-2')
instructorAndTimeArr = []
for i in range(len(instructorAndTime)):
    #Append all text elemeents into array
    instructorAndTimeArr.append(instructorAndTime[i].get_text())
    #Delete new lines and spaces
    instructorAndTimeArr[i] = " ".join(instructorAndTimeArr[i].strip().split())
    #Rid of "Office Hours"
    instructorAndTimeArr[i] = instructorAndTimeArr[i].replace(" Office Hours", "")
    
timeArr = []
instructorArr = []
for i in range(0,len(instructorAndTimeArr),3):
    instructorArr.append(instructorAndTimeArr[i])
    timeArr.append(instructorAndTimeArr[i+1])

#### Read the course schedule into a Dataframe

In [310]:
df = pd.DataFrame()
df['Number-Section'] = numbersectionarr
df['Name'] = coursenamearr
df['Instructor'] = instructorArr
df['Time'] = timeArr
df

Number-Section                                               Name  \
0     CSCI-GA1144-001                                            PAC II*   
1     CSCI-GA1144-002                                  PAC II Recitation   
2     CSCI-GA1170-001                             Fundamental Algorithms   
3     CSCI-GA1170-002                  Fundamental Algorithms Recitation   
4     CSCI-GA1170-003                  Fundamental Algorithms Recitation   
5     CSCI-GA1180-001        Mathematical Techniques For CS Applications   
6     CSCI-GA2110-001                              Programming Languages   
7     CSCI-GA2110-002                   Programming Languages Recitation   
8     CSCI-GA2110-003                   Programming Languages Recitation   
9     CSCI-GA2110-011                              Programming Languages   
10    CSCI-GA2110-012                   Programming Languages Recitation   
11    CSCI-GA2130-001                              Compiler Construction   
12    CSCI-GA2250-001                                  Operating Systems   
13    CSCI-GA2250-002                                  Operating Systems   
14    CSCI-GA2262-001                   Data Communications and Networks   
15    CSCI-GA2270-001                                  Computer Graphics   
16    CSCI-GA2421-001                               Numerical Methods II   
17    CSCI-GA2433-001                                   Database Systems   
18    CSCI-GA2433-011                                   Database Systems   
19    CSCI-GA2437-001                   Big Data Application Development   
20    CSCI-GA2440-001                               Software Engineering   
21    CSCI-GA2520-001                         Bioinformatics and Genomes   
22    CSCI-GA2560-001                            Artificial Intelligence   
23    CSCI-GA2565-001                                   Machine Learning   
24    CSCI-GA2572-001                                      Deep Learning   
25    CSCI-GA2572-002                                  Deep Learning Lab   
26    CSCI-GA2620-001                        Networks and Mobile Systems   
27    CSCI-GA2820-001                     DevOps and Agile Methodologies   
28    CSCI-GA2820-002                 DevOps and Agile Methodologies Lab   
29    CSCI-GA2945-002  Adv Topics Num. Analysis: High Performance Com...   
30    CSCI-GA2945-003  Adv Topics Num. Analysis: The Immersed Boundar...   
31    CSCI-GA2945-004  Adv Topics Num. Analysis: Stochastic Modeling ...   
32    CSCI-GA3033-004  Special Topics: Statistical Natural Language P...   
33    CSCI-GA3033-026                    Special Topics: Cloud Computing   
34    CSCI-GA3033-029                    Special Topics: Social Networks   
35    CSCI-GA3033-034  Special Topics: Multicore Processors: Architec...   
36    CSCI-GA3033-052          Special Topics: Advanced Machine Learning   
37    CSCI-GA3033-059                   Special Topics: Big Data Science   
38    CSCI-GA3033-071                 Special Topics: Geometric Modeling   
39    CSCI-GA3033-074        Special Topics: Practical Computer Security   
40    CSCI-GA3033-076      Special Topics: Vision Meets Machine Learning   
41    CSCI-GA3033-077            Special Topics: Big Data and ML Systems   
42    CSCI-GA3033-078  Special Topics: Cryptocurrencies and Decentral...   
43    CSCI-GA3033-079       Special Topics: Mathematics of Deep Learning   
44    CSCI-GA3033-085         Special Topics: Cloud and Machine Learning   
45    CSCI-GA3033-091  Special Topics: Introduction to Deep Learning ...   
46    CSCI-GA3033-117  Special Topics: Design and Analysis of Algorit...   
47    CSCI-GA3033-256                   Special Topics: Machine Learning   
48    CSCI-GA3033-991        Special Topics: Machine Learning Recitation   
49    CSCI-GA3033-992        Special Topics: Machine Learning Recitation   
50    CSCI-GA3033-993        Special Topics: Machine Learning Recitation   
51    CSCI-GA3140-001                            Abstra

#### Break apart Number-Section into Number and Section

In [311]:
df['Number'] = df['Number-Section'].str.extract('(.*-.*)-')
df['Section'] = df['Number-Section'].str.extract('.*-.*-(\d*)')
df = df.drop(['Number-Section'],axis = 1)

#### Show

In [312]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        152 non-null    object
 1   Instructor  152 non-null    object
 2   Time        152 non-null    object
 3   Number      152 non-null    object
 4   Section     152 non-null    object
dtypes: object(5)
memory usage: 6.1+ KB


In [313]:
df.head()

Name  \
0                            PAC II*   
1                  PAC II Recitation   
2             Fundamental Algorithms   
3  Fundamental Algorithms Recitation   
4  Fundamental Algorithms Recitation   

                                   Instructor           Time       Number  \
0                              Mohamed Zahran  T 6:00-8:30PM  CSCI-GA1144   
1                        Gurkirat Singh Bajwa  R 7:10-8:00PM  CSCI-GA1144   
2                              Yevgeniy Dodis  T 7:10-9:00PM  CSCI-GA1170   
3  Alex Bienstock Charles Peyser Fengyuan Liu  R 8:10-9:00PM  CSCI-GA1170   
4                          Harish Karthikeyan  R 8:10-9:00PM  CSCI-GA1170   

  Section  
0     001  
1     002  
2     001  
3     002  
4     003

In [314]:
df.tail()

Name        Instructor  \
147        Special Topics: Natural Language Processing       Adam Meyers   
148  Special Topics: Agile Software Development and...    Amos Bloomberg   
149        Special Topics: Algorithmic Problem Solving  Joanna Klukowska   
150        Special Topics: Algorithmic Problem Solving      Kunal Khatri   
151        Special Topics: Algorithmic Problem Solving    Samasth Ananda   

                 Time       Number Section  
147   TR 9:30-10:45AM  CSCI-UA0480     057  
148    MW 3:30-4:45PM  CSCI-UA0480     069  
149  MW 11:00-12:15PM  CSCI-UA0480     521  
150     W 3:30-4:45PM  CSCI-UA0480     522  
151     W 3:30-4:45PM  CSCI-UA0480     523

In [315]:
df.sample(5)

Name        Instructor  \
73   Intro To Computer Programming (No Prior Experi...  Hasan Aljabbouli   
102                       Data Structures - Recitation         CANCELLED   
140                              Theory Of Computation      Richard Cole   
101                                    Data Structures       Anasse Bari   
81   Intro to Computer Programming (Limited Prior E...  Hasan Aljabbouli   

                 Time       Number Section  
73    MW 12:30-1:45PM  CSCI-UA0002     005  
102                 -  CSCI-UA0102     004  
140   MW 12:30-1:45PM  CSCI-UA0453     001  
101   TR 9:30-10:45AM  CSCI-UA0102     003  
81   TR 11:00-12:15PM  CSCI-UA0003     001

## 2. Read the 2021 course catalog into a Dataframe

In [316]:
#Read the whole html
coursecat = BeautifulSoup(open("coursecat.htm", encoding = 'utf-8'), 'html.parser')
#The schedule information are all inside .schedule-listing class
coursecat1 = coursecat.select('.courses-listing')[0]


#### Parse number into array

In [317]:
#Count number of 'li', which are lists of individual classes
numcount = coursecat1.select('li')
numberarr = []
#Append all numbersection into array
for i in range(len(numcount)):
    numberarr.append(coursecat1.select('li')[i]['id'])
#Clean into appropriate format
for i in range(len(numberarr)):
    numberarr[i] = numberarr[i].upper()

#### Parse Prerequesites into array

In [318]:
#All prerequisites are inside class .expandable
prerequisites = coursecat1.select('.bold')

#Array to store prerequisites
prereqarr = []
counter = 0
#Append while removing newlines, space
for i in range(2, len(prerequisites), 3):
    prereqarr.append(prerequisites[i].get_text())
    prereqarr[counter] = " ".join(prereqarr[counter].strip().split())
    counter += 1

#### Parse Points to array

In [319]:
pointarr = []
counter = 0
for i in range(1, len(prerequisites), 3):
    pointarr.append(prerequisites[i].get_text())
    pointarr[counter] = pointarr[counter].strip().split()[0]
    counter += 1

#### Read arrays into dataframe

In [320]:
df1 = pd.DataFrame(columns = ['Number', 'Prereqs', 'Points'])
df1['Number'] = numberarr
df1['Prereqs'] = prereqarr
df1['Points'] = pointarr

#### Show

In [321]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Number   100 non-null    object
 1   Prereqs  100 non-null    object
 2   Points   100 non-null    object
dtypes: object(3)
memory usage: 2.5+ KB


In [322]:
df1.head()

Number                                            Prereqs Points
0  CSCI-GA1133                                Prerequisites: None      4
1  CSCI-GA1144  Prerequisites: CSCI-GA 1133 or departmental pe...      4
2  CSCI-GA1170  Prerequisites: At least one year of experience...      3
3  CSCI-GA1180                                Prerequisites: None      3
4  CSCI-GA2110  Prerequisites: Students taking this class shou...      3

In [323]:
df1.tail()

Number                                            Prereqs Points
95   CSCI-UA0897  Prerequisites: Restricted to declared computer...      1
96   CSCI-UA0898  Prerequisites: Restricted to declared computer...      1
97   CSCI-UA0997  Prerequisites: Permission of the department. D...      1
98   CSCI-UA0998  Prerequisites: Permission of the department. D...      1
99  FRSEM-UA0597  Prerequisites: Some programming experience in ...      4

In [324]:
df1.sample(5)

Number                                            Prereqs Points
56  CSCI-GA3812  Prerequisites: For MS in IS students: Successf...      3
2   CSCI-GA1170  Prerequisites: At least one year of experience...      3
5   CSCI-GA2112  Prerequisites: Multivariate calculus and linea...      3
10  CSCI-GA2262  Prerequisites: Students must have a working kn...      3
60  CSCI-GA3860  Prerequisites: Permission of the thesis advise...      1

## 3. Put together both DataFrames

In [325]:
df2 = pd.merge(df,df1, on='Number', how='left')
pd.set_option('display.max_rows', 200)
df2

Name  \
0                                              PAC II*   
1                                    PAC II Recitation   
2                               Fundamental Algorithms   
3                    Fundamental Algorithms Recitation   
4                    Fundamental Algorithms Recitation   
5          Mathematical Techniques For CS Applications   
6                                Programming Languages   
7                     Programming Languages Recitation   
8                     Programming Languages Recitation   
9                                Programming Languages   
10                    Programming Languages Recitation   
11                               Compiler Construction   
12                                   Operating Systems   
13                                   Operating Systems   
14                    Data Communications and Networks   
15                                   Computer Graphics   
16                                Numerical Methods II   
17                                    Database Systems   
18                                    Database Systems   
19                    Big Data Application Development   
20                                Software Engineering   
21                          Bioinformatics and Genomes   
22                             Artificial Intelligence   
23                                    Machine Learning   
24                                       Deep Learning   
25                                   Deep Learning Lab   
26                         Networks and Mobile Systems   
27                      DevOps and Agile Methodologies   
28                  DevOps and Agile Methodologies Lab   
29   Adv Topics Num. Analysis: High Performance Com...   
30   Adv Topics Num. Analysis: The Immersed Boundar...   
31   Adv Topics Num. Analysis: Stochastic Modeling ...   
32   Special Topics: Statistical Natural Language P...   
33                     Special Topics: Cloud Computing   
34                     Special Topics: Social Networks   
35   Special Topics: Multicore Processors: Architec...   
36           Special Topics: Advanced Machine Learning   
37                    Special Topics: Big Data Science   
38                  Special Topics: Geometric Modeling   
39         Special Topics: Practical Computer Security   
40       Special Topics: Vision Meets Machine Learning   
41             Special Topics: Big Data and ML Systems   
42   Special Topics: Cryptocurrencies and Decentral...   
43        Special Topics: Mathematics of Deep Learning   
44          Special Topics: Cloud and Machine Learning   
45   Special Topics: Introduction to Deep Learning ...   
46   Special Topics: Design and Analysis of Algorit...   
47                    Special Topics: Machine Learning   
48         Special Topics: Machine Learning Recitation   
49         Special Topics: Machine Learning Recitation   
50         Special Topics: Machine Learning Recitation   
51                             Abstract Interpretation   
52                   Information Technology Projects *   
53                                        Adv Lab MS *   
54                                        Adv Lab MS *   
55                                        Adv Lab MS *   
56                                        Adv Lab MS *   
57                                      Adv Lab PhD **   
58                                      Adv Lab PhD **   
59                                      Adv Lab PhD **   
60                                      Adv Lab PhD **   
61                          Master's Thesis Research *   
62                         PhD Seminar: Cryptography**   
63                       PhD Seminar: Formal Methods**   
64                              PhD Thesis Research **   
65                  Internship In Computer Science MS*   
66                 Internship in Computer Science PhD*   
67          Maintenance of Matriculation (MS Students)   
68   Maintenance of Matriculation (non-supported

## 4. Conclusion

#### Anomalies, discrenpancies, or unexpected data
- Points: Some had empty points so my parser had an error and gave "Points. " instead of the extra number
- Unequal number of courses between catalog and schedule: The catalog listed 100 courses while the schedule listed 152

#### How I may fix them

There is no real need to fix the unequal number of courses and catalog, since the merge already marks the missing data points as NaN. The Points part may get a better sophisticated parser that finds the string "Points." and writes the split() array of the index before it.

#### How (how = left) affects this dataframe
The second dataframe is appended to the right of the first dataframe with only the values it is missing